misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt

```diff
@@ -65,38 +68,56 @@ internal class RealTransacter private constructor(
       threadLatestSession = ThreadLocal(),
       options = TransacterOptions(),
       queryTracingListener = queryTracingListener,
-      tracer = tracer
-  )
+      executorServiceFactory = executorServiceFactory,
+      tracer = tracer,
+      shardListFetcher = ShardListFetcher()
+  ) {
+    shardListFetcher.init(this, config, executorServiceFactory)
+  }
 
   override fun config(): DataSourceConfig = config
 
-  private val shardListFetcher: ExecutorService by lazy {
-      Executors.newSingleThreadExecutor(
-          ThreadFactoryBuilder().setNameFormat("shard-list-fetcher-%d").build())
-  }
-
-  private val shardList = Suppliers.memoizeWithExpiration({
-    if (!config.type.isVitess) {
-      CompletableFuture.completedFuture(SINGLE_SHARD_SET)
-    } else {
-      // Needs to be fetched on a separate thread to avoid nested transactions
-      shardListFetcher.submit(Callable<Set<Shard>> {
-        transaction { session ->
-          session.useConnection { connection ->
-            connection.createStatement().use { s ->
-              val shards = s.executeQuery("SHOW VITESS_SHARDS")
-                  .map { rs -> Shard.parse(rs.getString(1)) }
-                  .toSet()
-              if (shards.isEmpty()) {
-                throw SQLRecoverableException("Failed to load list of shards")
+  /**
+   * Uses a dedicated thread to query Vitess for the database's current shards. This caches the
+   * set of shards for 5 minutes.
+   */
+  class ShardListFetcher {
```

Comment:
@jjestrel @tirsen FYI!

Reply:
nit: should we just have two and bind the one we need?

Reply:
Yeah, good idea. Coming in follow-up.